src/components/resilience_component.cpp

```diff
@@ -128,126 +125,100 @@ std::size_t resilience::protectee_rank_ = 0;
         failure_detection_loop_async();
     }
 
-    void resilience::global_w_exec_finish_wrapper(work_item const& w)
-    {
-        allscale::resilience::get().w_exec_finish_wrapper(w);
-    }
-
-    // equiv. to taskAcquired in prototype
-    void resilience::w_exec_start_wrapper(work_item const& w) {
-        if (get_running_ranks() < 2 || env_resilience_disabled) 
-            return;
-
-        if (w.id().depth() != get_cp_granularity()) return;
-
-        //@ToDo: do I really need to block (via get) here?
-        if (get_running_ranks() > 1) {
-            hpx::async<remote_backup_action>(guard_, w).get();
-	    std::unique_lock<mutex_type> lock(backup_mutex_);
-	    //thread_safe_printer("local_backups_["+w.id().name()+"]\n");
-	    local_backups_[w.id().name()] = w;
+    void resilience::reschedule_dispatched_to_dead(size_t dead_rank, size_t token) {
+        std::multimap<size_t, work_item > delegated_items_copy;
+        {
+            std::unique_lock<mutex_type> lock(delegated_items_mutex_);
+            std::swap(delegated_items_, delegated_items_copy);
+        }
+        {
+            std::unique_lock<mutex_type> lock(running_ranks_mutex_);
+            thread_safe_printer("rank["+std::to_string(dead_rank)+"]=false");
+            rank_running_[dead_rank] = false;
+            auto protectee_locality = get_locality_from_id(protectee_);
+            auto it = std::find(localities.begin(), localities.end(), protectee_locality);
+            localities.erase(it);
+            num_localities--;
+        }
+        auto range = delegated_items_copy.equal_range(dead_rank);
+        for (auto it = range.first; it != range.second; it++) {
+            auto w = it->second;
+            auto id = w.id();
+            thread_safe_printer("Reschedule delegated item:"+id.name());
+            allscale::scheduler::schedule(std::move(w), id);
+        }
+        delegated_items_copy.erase(dead_rank);
+        {
+            std::unique_lock<mutex_type> lock(delegated_items_mutex_);
+            std::swap(delegated_items_, delegated_items_copy);
```

Comment:
This might loose delegated items that have been enqueued during the restore operation.